programs/README.md

```diff
@@ -0,0 +1,123 @@
+
+* [`random/gen_random_ctr_drbg.c`](random/gen_random_ctr_drbg.c): illustrates using the default entropy sources to seed a pseudorandom generator, and using the resulting random generator to generate random data.
+
+* [`random/gen_random_havege.c`](random/gen_random_havege.c): illustrates the HAVEGE entropy collector.
+
+## SSL/TLS examples
+
+### SSL/TLS sample applications
+
+* [`ssl/dtls_client.c`](ssl/dtls_client.c): a simple DTLS client program which sends one datagram to the server and reads one datagram in response.
+
+* [`ssl/dtls_server.c`](ssl/dtls_server.c): a simple DTLS server program which expects one datagram from the client and writes one datagram in response. This program supports DTLS cookies for hello verification.
+
+* [`ssl/mini_client.c`](ssl/mini_client.c): a minimalistic SSL client which sends a short string and disconnects. This is intended more as a benchmark; for a better example of a typical TLS client, see `ssl/ssl_client1.c`.
+
+* [`ssl/ssl_client1.c`](ssl/ssl_client1.c): a simple HTTPS client that sends a fixed request and displays the response.
+
+* [`ssl/ssl_fork_server.c`](ssl/ssl_fork_server.c): a simple HTTPS server using one process per client to send a fixed response. This program requires a Unix/POSIX environment implementing the `fork` system call.
+
+* [`ssl/ssl_mail_client.c`](ssl/ssl_mail_client.c): a simple SMTP-over-TLS or SMTP-STARTTLS client. This client sends an email with a fixed content.
+
+* [`ssl/ssl_pthread_server.c`](ssl/ssl_pthread_server.c): a simple HTTPS server using one thread per client to send a fixed response. This program requires a the pthread library.
+
+* [`ssl/ssl_server.c`](ssl/ssl_server.c): a simple HTTPS server that sends a fixed response. This server serves a single client at a time.
+
+### SSL/TLS feature demonstrators
+
+Note: unlike most of the other programs under the `programs/` directory, these two programs are not intended as a basis to start writing an application. They combine most of the features supported by the library, and most applications require only a few features. It is recommended to start with `ssl_client1.c` or `ssl_server.c`, and to look inside `ssl/ssl_client2.c` or `ssl/ssl_server2.c` to see how to use the specific features that your application needs.
+
+* [`ssl/ssl_client2.c`](ssl/ssl_client2.c): an HTTPS client that sends a fixed request and displays the response, with options to select TLS protocol features and Mbed TLS library features.
+
+* [`ssl/ssl_server2.c`](ssl/ssl_server2.c): an HTTPS server that sends a fixed response, with options to select TLS protocol features and Mbed TLS library features.
+
+These programs have options to trigger certain behaviors (e.g. reconnection, renegotiation) so the `ssl_server2` program can be useful to test features in your TLS client and the `ssl_client2` program can be useful to test features in your TLS server.
```

Comment:
Comma needed before 'so,' but sentence is really long and the wording a bit awkward - I would change to: "These programs provide options for triggering certain behaviors (e.g. reconnection, renegotiation). So, the `ssl_server2` program can be useful for testing features in you TLS client, and the `ssl_client2` program can be useful for testing features in your TLS server."

Reply:
I reworded the paragraph.

Reply:
Still need a few changes to the reworded paragraph:
"In addition to providing options for testing client-side features, the `ssl_client2` program has options to exercise certain behaviors in the server (for example, to select ciphersuites, or to force a renegotiation), which are useful to test the corresponding features in a TLS server. Likewise, `ssl_server2` has options to activate certain behaviors that are useful to test a TLS client."

1) "has options to exercise" doesn't sound right, and "exercise certain behaviors" also sounds off.  It would be better to say "allows you to trigger certain behaviours in the server..." (you had 'trigger certain behaviors' originally, and I think that sounds better than 'exercise').

2) Also, we can't use () with a long strand of words inside beginning with "which" - i.e. elaborating in ().  So, the sentences would need to be: "In addition to providing options for testing client-side features, the `ssl_client2` program allows you to trigger certain behaviours in the server, for example, to select ciphersuites or to force a renegotiation...
However, the sentence is still too long.  We have to make sure that sentences are short in the documentation because, otherwise, it is hard for non-native speakers to follow.  So, I would say:

"In addition to providing options for testing client-side features, the `ssl_client2` program allows you to trigger certain behaviours in the server, for example, to select ciphersuites or to force a renegotiation. These are useful for testing the corresponding features in a TLS server."
(note: "to test" also needs to be changed to "for testing" above)

3) The last sentence: "has options" should be "gives you the option".  `ssl_server2` is a program so it should have the definite article before it and 'program' after it (you had "the `ssl_server2` program before and that sounds better).  'to test' needs to be 'for testing.'  So:

"Likewise, the `ssl_server2` program gives you the option to activate certain behaviors that are useful for testing a TLS client."

Reply:
I'm splitting that runaway sentence. However, “give the option” doesn't work here because this is this not the vernacular meaning of the word “option”. The options are [command line options](https://en.wikipedia.org/wiki/Command-line_interface#Command-line_option), and that's a countable usage.

Reply:
The new sentences look great.